src/viewport-observer.js

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {isIframed} from './dom';
+
+/**
+ * Returns an IntersectionObserver tracking the Viewport.
+ * Only use this if x-origin iframe rootMargin support is considered nice-to-have,
+ * since if not supported this InOb will fallback to one without rootMargin.
+ *
+ * - If iframed: rootMargin is ignored unless natively supported (Chrome 81+).
+ * - If not iframed: all features work properly in both polyfill and built-in.
+ *
+ * @param {function(!Array<!IntersectionObserverEntry>)} ioCallback
+ * @param {!Window} win
+ * @param {string=} threshold
+ *
+ * @return {!IntersectionObserver}
+ */
+export function getViewportObserver(ioCallback, win, threshold) {
+  const iframed = isIframed(win);
+  const root = /** @type {?Element} */ (iframed
+    ? /** @type {*} */ (win.document)
+    : null);
+
+  // TODO(amphtml): Determine if rootMargin is still necessary.
+  const rootMargin = '25%';
+
+  try {
+    return new win.IntersectionObserver(ioCallback, {
+      root,
+      rootMargin,
+      threshold,
+    });
+  } catch (e) {
+    return new win.IntersectionObserver(ioCallback, {rootMargin, threshold});
+  }
+}
+
+/** @type {Map<!Window, !IntersectionObserver>} */
```

Comment:
TBH, it's bugging me a bit, that some user would only want to use `getViewportObserver`, but would get these two side-effects. Not a huge deal, of course. But do we really need to export `getViewportObserver`? Can it stay private to this module?

Reply:
I see that you are using it to create an independent observer for videos. I guess it's ok. But consider it and see if you're ok with this.

Reply:
I was thinking that `video-manager-impl` would use `getViewportObserver` and the rest would directly use `observe`, `unobserve`.

Reply:
Yep. I saw that one later. I'm ok to keep as is. A huge improvement anyway. Not sure of closure is smart enough to remove such side-effects, but it's a very low cost.

Reply:
you're right of course. I wish i knew good patterns for writing code that has no module-level side effects

Reply:
I think you'd split it into two modules classically. But I think in this case may be not worth it.